proposals/nnnn-variadic-types.md

```diff
@@ -0,0 +1,239 @@
+
+```swift
+struct V<each T> {}
+
+let x: V<_> = V<Int>()  // okay
+let x: V<_, _> = V<Int, String>()  // okay
+let x: V<_> = V<Int, String>()  // error
+```
+
+### Stored properties
+
+In a variadic type, the type of a stored property can contain pack expansion types. The type of a stored property cannot _itself_ be a pack expansion type. Stored properties are limited to having pack expansions nested inside tuple types, function types and other named variadic types:
+
+```swift
+struct S<each T> {
+  var a: (repeat each Array<T>)
+  var b: (repeat each T) -> (Int)
+  var c: Other<repeat each T>
+}
+```
+This is in contrast with the parameters of generic function declarations, which can have a pack expansion type. A future proposal might lift this restriction and introduce true "stored property packs":
+
+```swift
+struct S<each T> {
+  var a: repeat each Array<T>
+}
+```
+
+### Requirements
+
+The behavior of generic requirements on type parameter packs is mostly unchanged between generic functions and generic types. However, allowing types to abstract over parameter packs introduces _requirement inference_ of [generic requirement expansions](https://github.com/apple/swift-evolution/blob/main/proposals/0393-parameter-packs.md#generic-requirements). Requirement expansion inference follows these rules:
+
+1. If a generic type that imposes an inferred scalar requirement is applied to a pack element inside a pack expansion, the inferred requirement is a requirement expansion.
+2. If a generic type imposes an inferred requirement expansion, the requirement is expanded for each of the concrete generic arguments.
+3. If a generic type that imposes an inferred requirement expansion is applied to a pack element inside a pack expansion:
+
+    1. The inferred requirement is invalid if it contains multiple pack elements captured by expansions at different depths.
+    2. Otherwise, the nested requirement expansion is semantically equivalent to the innermost requirement expansion.
+  
+  The below code demonstrates each of the above rules:
+
+  ```swift
+  protocol P {
+      associatedtype A
+  }
+  struct ImposeRequirement<T> where T: P {}
+  struct ImposeRepeatedRequirement<each T> where repeat each T: P {}
+  struct ImposeRepeatedSameType<T1: P, each T2> where repeat T1.A == each T2 {}
+  
+  // Infers 'repeat each U: P'
+  func demostrate1<each U>(_: repeat ImposeRequirement<each U>)
+  
+  // Infers 'Int: P, V: P, repeat each U: P'
+  func demostrate2<each U, V>(_: ImposeRepeatedRequirement<Int, V, repeat each U>)
+  
+  // Error. Would attempt to infer 'repeat <U' = each U> repeat <V' = each V> U'.A == V' which is not a supported requirement in the language
+  func demonstrate3a<each U, each V>(_: repeat ImposeRepeatedSameType<each U, repeat each V>))
+  
+  // Infers 'Int: P, repeat each V: P'
+  func demostrate3b<each U, each V>(_: repeat (each U, ImposeRequirementExpansion<Int, repeat each V>))
```

Comment:
Is this missing a type definition above for `ImposeRequirementExpansion` or should it reuse `ImposeRepeatedRequirement`?

Reply:
Oops, I renamed `ImposeRequirementExpansion` to `ImposeRepeatedRequirement` and forgot to fix this line. It should be `ImposeRepeatedRequirement`